src/Orleans.Serialization.MessagePack/MessagePackCodec.cs

```diff
@@ -68,14 +69,15 @@ void IFieldCodec.WriteField<TBufferWriter>(ref Writer<TBufferWriter> writer, uin
         // Write the type name
         ReferenceCodec.MarkValueField(writer.Session);
         writer.WriteFieldHeaderExpected(0, WireType.LengthPrefixed);
-        writer.Session.TypeCodec.WriteLengthPrefixed(ref writer, value.GetType());
+        var serializedType = expectedType != ObjectType ? expectedType ?? value.GetType() : value.GetType();
```

Comment:
This is the only part I wasn't too sure about, when telling MessagePack to deserialize into an `Object` type, it creates nested arrays of type `Object[]` which fails the unit tests.

Reply:
Using `value.GetType()` passes all tests, but MessagePack supports `IMessagePackFormatter<T>` with custom logic for serializing type `T`, using `value.GetType()` may produce different results to `expectedType`.
For example using `value.GetType()` when serializing `IMyMessagePackUnion` will add extra information to the header of the serialized array about the union type `Key`, whether that really makes a difference in the context of Orleans, I'm not too sure.
But, I agree that the above code produces different results for when Orleans requires `Object` vs `expectedType` so it already has that discrepancy.
May I ask, when does Orleans set expectedType as `Object`?

Reply:
I agree, I can't think of a practical use case where IMessagePackFormatter<T>.Serialize is changing the value of the object being serialized. I've updated the PR to always use `value.GetType()`